### DISCUSSION TITLE
Creating a base class for models

### DISCUSSION THREAD

**Discussion by nicreichert** [prisma/prisma#23222] [Q&A]
### Question

Hey! 👋 

I'm trying to create a base class to create basic and recurrent requests to the models that can be extended. The issue that comes up is somewhat recurrent in the issues section, but none of the proposed solutions worked for my case.

here's an example of the base class:

```ts
type ModelName = keyof typeof Prisma.ModelName

export class BaseModel<T extends ModelName> {
  protected model: PrismaClient[T]

  constructor(modelName: T) {
    this.model = db[modelName]
  }

  findUnique = <Args = Prisma.Args<T, 'findUnique'>, Result = Prisma.Result<T, Args, 'findUnique'>>(
    args: Args,
  ): Result => this.model.findUnique(args)

  findMany = <Args = Prisma.Args<T, 'findMany'>, Result = Prisma.Result<T, Args, 'findMany'>>(
    args: Args,
  ): Result => this.model.findMany(args)

  create = <Args = Prisma.Args<T, 'create'>, Result = Prisma.Result<T, Args, 'create'>>(
    args: Args,
  ): Result => this.model.create(args)

  update = <Args = Prisma.Args<T, 'update'>, Result = Prisma.Result<T, Args, 'update'>>(
    args: Args,
  ): Result => this.model.update(args)

  delete = <Args = Prisma.Args<T, 'delete'>, Result = Prisma.Result<T, Args, 'delete'>>(
    args: Args,
  ): Result => this.model.delete(args)
}
```

The problems that I'm facing are:

1. Even though I'm explicitly setting the type of the returns and params, when I extend it in a consumer class, they are typed with `any` (for `args`) and `{}` (for `result`).

Here's an example of a model implementation:

```ts
class UserModel extends BaseModel<'user'> {
  constructor() {
    super('user')
  }

  getByEmail = async (email: string) => this.findUnique({ where: { email } })

  getById = async (id: USER_ID) => this.findUnique({ where: { id } })

  edit = async (id: USER_ID, data: p.Prisma.calendarUpdateInput) =>
    this.update({ where: { id }, data })

  remove = async (id: USER_ID) => this.delete({ where: { id } })
}
```

Note: If instead of using `this.findUnique` (or any other of the base methods) I use `this.model.findUnique`, the types are working as expected.

2. The base class has a type issue that I've seen around the issues, but none of the proposed fixes helped my case.

![Screenshot 2024-02-21 at 09 45 47](https://github.com/prisma/prisma/assets/4841711/d1f80369-91f7-4d8a-a158-3226d8610a67)

I'd really appreciate some pointers! 💪 

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: macOS
- Database: PostgreSQL
- Node.js version: 18.6.0
- prisma version: 5.9.1
- @prisma/client version: 5.9.1

**Comment by ludralph** [comment]
Hi @nicreichert 👋 

Thank you for raising this question.

Prisma relies on type inference based on your schema definitions. When you introduce a base class, the type information gets lost during method calls like `this.findUnique()`. The compiler can't infer the specific model type ('user') from the base class definition. Instead of using base class methods, directly call the specific model methods on `this.model` just like you have suggested. This leverages Prisma's type inference for accurate typing.
```ts
class UserModel {
  constructor() {}

  getByEmail = async (email: string) => this.model.user.findUnique({ where: { email } })

  getById = async (id: USER_ID) => this.model.user.findUnique({ where: { id } })

  // ... other methods
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
